app/lib/csv_importer.rb

```diff
@@ -10,29 +10,49 @@ class CsvImporter
 
   class InvalidCategoryError < StandardError; end;
 
-  def self.import(filename, category_name, processed_file_id)
-    importer = new(filename, category_name, processed_file_id)
-    importer.process
-    importer.stats
-  end
-
   def initialize(filename, category_name, processed_file_id)
     @filename = filename
     @processed_file_id = processed_file_id
     @model = model_from_category(category_name)
-
     @stats = Hash.new(0)
+    @error_details = {}
     @stats[:shl_case_numbers] = Hash.new(0)
   end
 
+  def call
+    process
+  end
+
+  def errored?
+    !error_details.empty?
+  end
+
+  private
+
   def process
-    IOStreams.each_record(filename) do |csv_row|
-      csv_row[:processed_file_id] = processed_file_id
-      csv_row[:raw] = false
-      record = model.create_from_csv_data(csv_row)
-      record.cleanse_data! if record.respond_to?(:cleanse_data!)
-      record.save
-      increment_stats(record)
+    row_number = 2 # assuming 1 is headers
+
+    model.transaction do
+      IOStreams.each_record(filename) do |csv_row|
+        csv_row[:processed_file_id] = processed_file_id
+        csv_row[:raw] = false
+        record = model.create_from_csv_data(csv_row)
+        record.cleanse_data! if record.respond_to?(:cleanse_data!)
+
+        if record.valid?
+          record.save
+          increment_stats(record)
+        else
+          error_details["row_number_#{row_number}"] = record.errors.details
+        end
```

Comment:
Thoughts on using a `begin/rescue` /w `save!` approach here and keeping all else the same? I've noticed that sometimes `create_from_csv_data` will raise an error if the DateTime is not parsed correctly. 

For example in `app/models/tagged_animal_assessment.rb:55` if the value of `measurement_date_str` is not valid it will trigger a raised error.

I discovered this while working on https://github.com/rubyforgood/abalone/issues/97

Reply:
I think that's the direction @wesleyeewong had gone earlier, but @corneliusellen pointed out how it would make it harder for users to understand the problem.

Reply:
@edwinthinks  ah I see what you mean, it fails before hitting the `valid?` check, on the record creation line `model.create_from_csv_data(csv_row)`

Error reproducible by changing `Measurement_date` in `spec/support/csv/Tagged_assessment_invalid_values.csv` to a non date format string.

I don't think `begin/rescue` would solve the issue tho. 🤔 And I'm not sure what the best solution would be.